### ISSUE TITLE
Improve UX and UI of "Sign an operation" pop-up

### ISSUE THREAD

**Issue by leoloco** [massalabs/station-massa-wallet#752]
## TL;DR
In Massa, the overall transaction cost is `fee + coins` where `fee` is the compensation paid to the validators for computational work and `coins` is the amount of coins locked for storage space on the ledger. I don't think that users understand this when they look at the "Sign an operation" pop-up. Moreover, in Massa's DApps, `maxGas` is set by the frontend. This should be set by the wallet.

- Suggestion (1): In the "Sign an operation" pop-up, I suggest that we hide the details (Coins,Fees,Max gas,Expiry) and instead provide a single `transaction cost` field that would equal `fee + coins`. There could be a dropdown that explains the fee structure for more advanced users. 
- Suggestion (2): I suggest that we add a layer to the wallet that will automatically evaluate and set the `maxGas` parameter by calling the readOnly method. 

## Context
As of today, the "Sign an operation" pop-up displays the following information :
- From
- To
- Operation Type
- Function called
- Coins
- Fees
- Max gas
- Expiry

The first four parameters are more than relevant to the user, however, I don't think that the same can be said about the following ones. On the basis that users are familiar with the EVM environment, I'm afraid that the actual meaning and implication of these parameters might be obscure to the user.

#### Transaction cost model in EVM chains
in EVM chains, the gas is a unit that represents computational work. This also applies in Massa. Both protocols start to diverge on the notion of gas price. In EVM chains the `transaction fee` is the produce of `gas used` and `gas price`. The `gas price` is then used by the validators to determine the profitability per unit of computational work of each transaction. Based on this, they can compute a block that maximizes their profits. 

#### Transaction cost model in Massa
In Massa, the notion of `gas price` doesn't exist. Instead, the user sets a `fee` and a `maxGas` parameter. The `fee`, expressed in MAS tokens, will be consumed by the block producer if the transaction is included in the block, no matter if the execution is successful or not. The `maxGas` parameter is an equivalent of `gasLimit` in EVM chains. But in Massa, the `maxGas` parameter is used by the validators  EVM `gasPrice` can computed by the validators by calculating the ratio between `maxGas` and `fee`. 

#### Difference between the two models
The difference between these models can be explained by the following fact; in EVM chains, when populating a block with transactions, validators are aware of the  `gas used` and the `gasPrice` associated with each transaction. 
In Massa, because of processing-time constraints, validators don't know in advance how much gas will be used by each transaction. Thus, they can only evaluate the profitability of each transaction, by computing the ratio between `fee` and `maxGas` provided by the user. 

#### The fee equilibrium
In Massa, if the `maxGas` parameter provided by the user isn't enough to cover the computational work of the transaction, the transaction will be reverted. On the other hand, if the user provides a `maxGas` parameter that is overwhelmingly high compared to the actual computational work, the transaction will seem less profitable for the validator, and thus, reduce the chances of the transaction being included. 
Thus, in Massa, the users should set `maxGas` sufficiently high to cover computational work but also sufficiently low to appear as a more profitable transaction. 

## Need 
Based on those observations, I think that the current user experience of signing an operation is highly misleading for users. Instead of seeing the transaction cost, which in my opinion is the information of importance, they see three different parameters that have overlapping meanings and are difficult to understand. (plus the expiry parameter that ended up here for whatever reason). 

#### Need 1: a simple UI that showcases the transaction cost
Instead of showing the four parameters that are currently shown, I think that we should provide a more understandable `transaction cost` parameter. This parameter could be the sum of `fee`+`coins`. We could also show something that reflects the ratio between `fee` and `maxGas` as an equivalent of the `gasPrice` parameters that EVM user are used to. The user should also have access to a breakdown of the fee structure through a "details" dropdown or something like that.

- [ ] Part of the need will be addressed here: https://github.com/massalabs/station-massa-wallet/issues/786

#### Need 2: an intermediate layer on the wallet that automatically evaluates maxGas and sets it by default
In Massa, the `maxGas` parameter is set by the DApp front-end and forwarded to the wallet. This results in developers setting overwhelmingly high `maxGas` for safety and the users having to sign those high numbers (see picture).  In my opinion, `maxGas` should be evaluated and set directly by the wallet. Web3 lib should still be able to set a default `maxGas` but this should be overwritten by the wallet (like in EVM chains with metamask for example). This can be done by executing a read-only SC call.
PS: we should emphasize the fact that the "coins" parameters can be refunded on storage space freeing.  

![image](https://github.com/massalabs/station-massa-wallet/assets/12166652/259f701b-3473-47dc-9de2-ceb8c00b4060)

**Comment by peterjah** [comment]
for 1) sure a share the point:
for 2) I don't think this is the wallet job to estimate theses value (maxGas and Coins).
As you said this is a readonly call, so wallet should not be used to this purpose (wallet is focusing on write operation)
I think what you describe should be done in Massa-web3.
BTW readOnlyCall does not work in buildnet (i think it has been fixed in testnet)

another thing you mentionned is taht the user should be able to modify values for op parameters: fee, coins and maxgas. I agree on that. but the calling dApp should always provide a default value (which will be used as it is most of the time)

**Comment by leoloco** [comment]
Yes ok, I understand the point of view. But anyway, I think we need a calibrator for DApps developer. The logic behind gas cost calibration is pretty complex, and it would be nice for developers to have a simple yet sufficiently efficient tool to avoid dealing with such complex problem.

**Comment by Thykof** [comment]
@leoloco @peterjah @longgamma-massa for an execute SC operation, we don't have coins parameter but max coins. I don't think we can say cost = max coins + fees. How to handle this? 

we can show to user:

```
max coins: 5 MAS
operation cost: 0.001 MAS
  details:
  fees: 0.001 MAS
```

**Comment by leoloco** [comment]
Well, it depends on the behavior of the execute SC operation. If the overhead storage cost is automatically reimbursed, then indeed, you'd have to separate max coins and op cost. If the overhead storage cost is not automatically reimbursed then you can merge the costs in the UI (and provide details in a dropdown).

**Comment by Thykof** [comment]
for transaction, just show the amount to send and able to edit the fees

**Comment by Thykof** [comment]
function to estimate the gas consumption of an operation in typescript:

```ts
 public async getDynamicGasCost(
        targetAddress: string,
        targetFunction: string,
        parameter: number[],
    ): Promise<bigint> {
        let val: bigint;
        try {
            const readOnlyCall = await this.client.smartContracts().readSmartContract({
                targetAddress: targetAddress,
                targetFunction: targetFunction,
                parameter,
                maxGas: BigInt(MAX_GAS),
            });
            val = BigInt(Math.min(Math.floor(readOnlyCall.info.gas_cost * 1.2), MAX_GAS));
            this.lastMaxGas = { ...this.lastMaxGas, [targetAddress + targetFunction]: val };
        } catch (err) {
            this.logger.error(
                `Failed to get dynamic gas cost for ${targetFunction} at ${targetAddress}. Using fallback value `,
                err,
            );
            if (this.lastMaxGas[targetAddress + targetFunction]) {
                val = this.lastMaxGas[targetAddress + targetFunction];
            } else {
                val = BigInt(MAX_GAS);
            }
        }
        return val;
    }
```

**Comment by leoloco** [comment]
In assembly script (so SC side), this method is suggested to allow for storage coins reimbursement 
```ts

export function refund(initialBalance: u64): void {

    // generateEvent(`initialBalance: ${initialBalance}`);
    const newBalance: u64 = balance();
    // generateEvent(`newBalance: ${newBalance}`);
    // Should never assert (or something is seriously wrong in the blockchain)
    assert(initialBalance >= newBalance, "Runtime error");
    let balanceDelta: u64 = initialBalance - newBalance;
    // generateEvent(`balanceDelta: ${balanceDelta}`);

    let transferredCoins = Context.transferredCoins();
    if (transferredCoins > 0) {
        // Only refund if caller has transferred too much coins (parameter coins of callSmartContract)
        let coinsToRefund = transferredCoins > balanceDelta ? transferredCoins - balanceDelta : 0;
        if (coinsToRefund > 0) {
            // generateEvent(`[refund] send back ${coinsToRefund} coins`);
            transferCoins(Context.caller(), coinsToRefund);
        }
    } else {
        // read only call - transferred coins is set to 0
        // to estimate gas cost & Storage cost
        // TEMP: need an event to retrieve gas cost
        let storageCost: u64 = balanceDelta;
        generateEvent(`Estimated storage cost: ${storageCost}`);
    }
}

```

**Comment by Thykof** [comment]
> In assembly script (so SC side), this method is suggested to allow for storage coins reimbursement

@leoloco how should we use this function? should we call it explicitly on certain condition or at the end of each function of the SC?

**Comment by leoloco** [comment]
I don't know, I think the whole middle layer for gas and storage cost calibration needs more conception. The code I provided above is just an idea.

**Comment by Thykof** [comment]
> Well, it depends on the behavior of the execute SC operation. If the differential between storage cost and max coins is automatically reimbursed, then indeed, you'd have to separate max coins and op cost. If the differential between storage cost and max coins is not automatically reimbursed then you can merge the costs in the UI (and provide details in a dropdown).

thanks to the understanding of https://docs.massa.net/docs/learn/operation-format-execution#operation-execution the coins of max_coins are not spent if the execution fails, but the fees are spent.

So we should not include max_coins into the operation cost for a user point of view.

This mean that this proposition is good: 

> we can show to user:
> 
> ```
> max coins: 5 MAS
> operation cost: 0.001 MAS
>   details:
>   fees: 0.001 MAS
> ```

**Comment by leoloco** [comment]
In my understanding of the documentation, if max_coins is set to 5MAS and the operation is successful then 5MAS will be spent. Thus why wouldn't you include them in the transaction cost ?

**Comment by Thykof** [comment]
I understand that the caller can spend up to max_coins during the execution of the operation. I understand that the bytecode of the operation can spend less than this parameter and in that case the amount spent will be less than the max_coins parameter. @massalabs/core-team could you help us understanding how execute SC works please?

**Comment by Eitu33** [comment]
> I understand that the bytecode of the operation can spend less than this parameter

@Thykof on what observations do you base this?

**Comment by Thykof** [comment]
> > I understand that the bytecode of the operation can spend less than this parameter
> 
> @Thykof on what observations do you base this?
> 
> Edit: nevermind I see what you mean now, I understand Aurélien has already answered you yes?

If the the caller can spend up to max_coins, it means it can spend less. Yes we have all information.

**Comment by SlnPons** [comment]
Done with https://github.com/massalabs/station-massa-wallet/issues/786
